### PR TITLE
calico-kube-controllers fix concurrent map writes issue

### PR DESCRIPTION
**Pull request by zamog**
## Description
Add mutex lock to the Set function to prevent a race condition that causes the process to panic with fatal error: concurrent map read and map write

## Related issues/PRs
fixes [8705](https://github.com/projectcalico/calico/issues/8705)

## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
TBD
```

## Reminder for the reviewer

Make sure that this PR has the correct labels and milestone set.

Every PR needs one `docs-*` label.

- `docs-pr-required`: This change requires a change to the documentation that has not been completed yet.
- `docs-completed`: This change has all necessary documentation completed.
- `docs-not-required`: This change has no user-facing impact and requires no docs.

Every PR needs one `release-note-*` label.

- `release-note-required`: This PR has user-facing changes. Most PRs should have this label.
- `release-note-not-required`: This PR has no user-facing changes.

Other optional labels:

- `cherry-pick-candidate`: This PR should be cherry-picked to an earlier release. For bug fixes only.
- `needs-operator-pr`: This PR is related to install and requires a corresponding change to the operator.

### CODE REVIEW

**Comment by aaaaaaaalex** [comment]
Thanks for the PR.

Based on your issue, I'm not 100% convinced this will fix the race. If the concurrent write is also occurring in the same `Set` method, then maybe, but, It seems from your logs of the issue that:

 - `reflect.DeepEqual()` being called on the cache **item** is happening concurrently to a *write* on that item.

- The item being mutated is a `WorkloadEndpointData`, particularly, it seems like a mutation on it's `labels` map may be the culprit. Therefor I'm guessing the panic arises from a write to this struct's `labels` while the reflect pkg reads them.

Do we know what component is *writing* to that `labels` map while `reflect` is reading it? And can we confidently say that synchronising calls of the `Set` method removes that concurrency?



On a side-note, our process would be to commit this fix to master branch, rather than directly to the release branch, to ensure we don't regress in future releases. We can then backport the master patch to 3.26.

File: kube-controllers/pkg/cache/cache.go [hunk 1]
```diff
@@ -126,7 +126,9 @@ func (c *calicoCache) Set(key string, newObj interface{}) {
 	if reflect.TypeOf(newObj) != c.ObjectType {
 		c.log.Fatalf("Wrong object type received to store in cache. Expected: %s, Found: %s", c.ObjectType, reflect.TypeOf(newObj))
 	}
-
+	// lock the cache
```
Comment (caseydavenport) [inline kube-controllers/pkg/cache/cache.go:129]: Something isn't quite adding up to me here - the underlying map is within the `threadSafeCache` instance, which IIUC should already handle this. 

Do you know why we need this second mutex?

**Comment by matthewdupre** [comment]
@zamog I think we'd be open to adding any necessary locking here, but we need to understand how the concurrent writes happen: adding locks without understanding why is not generally safe.  If you can come up with a UT or description of how the concurrent writes happen then we can make a fix.

**Pull request closed without merging**